### ISSUE TITLE
org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to exprId on unresolved object, tree: 'rel0.p

### ISSUE THREAD

**Issue by pj001** [SANSA-Stack/SANSA-Stack#124]
org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to exprId on unresolved object, tree: 'rel0.p

**Comment by pj001** [comment]
'Project ['rel0.s, 'rel1.o, 'rel0.o]
+- 'Filter ('rel1.p = http://www.w3.org/2000/01/rdf-schema#subPropertyOf)
   +- 'Join Inner, ('rel0.p = 'rel1.s)
      :- 'SubqueryAlias rel1
      :  +- 'UnresolvedRelation [TRIPLES]
      +- 'SubqueryAlias rel0
         +- 'UnresolvedRelation [TRIPLES]

**Comment by LorenzBuehmann** [comment]
would be helpful if you provide code, data and SPARQL query

A minimal example makes support way easier - so far, all I can deduce is that you most likely run a SPARQL query like
```sparql
select * where {?s rdfs:subPropertyOf ?o . ... }
```
but not with which SANSA version, nor do I know how you run the query and which data.

Thank you

**Comment by pj001** [comment]
> 如果您提供代码，数据和SPARQL查询，将很有帮助
> 
> 一个最小的示例使支持方式更容易-到目前为止，我可以推断出的是，您很可能会运行SPARQL查询，例如
> 
> ```sparql
> 选择*其中{ ？s rdfs：subPropertyOf ？o。...}
> ```
> 
> 但不知道使用哪个SANSA版本，也不知道如何运行查询和哪些数据。
> 
> 谢谢

https://github.com/SANSA-Stack/SANSA-Stack/issues/124#issuecomment-749486340

**Comment by pj001** [comment]
I tried to run spark inference, but I didn't know what file was read in the example.... So I haven't figured out how the program works. I hope to provide a clear use case

**Comment by pj001** [comment]
I tried to run spark inference, but I didn't know what file was read in the example.... So I haven't figured out how the program works. I hope to provide a clear use case

**Comment by LorenzBuehmann** [comment]
can you be more specific please.

What actions exactly did you do? We have to see code and/or how you run whatever you did.
And with which SANSA Stack version?

**Comment by pj001** [comment]
I don't know how this is reasoned and how the file is read.&nbsp;I read the source code, it seems that there is no Tu// load triples from disk
val graph = RDFGraphLoader.loadFromDisk(spark, input, parallelism)
println(s"|G| = ${graph.size()}")

// create reasoner
val reasoner = profile match {
  case TRANSITIVE =&gt; new TransitiveReasoner(spark.sparkContext, properties, parallelism)
  case RDFS =&gt; new ForwardRuleReasonerRDFS(spark.sparkContext, parallelism)
  case RDFS_SIMPLE =&gt;
    val r = new ForwardRuleReasonerRDFS(spark.sparkContext, parallelism)
    r.level = RDFSLevel.SIMPLE
    r
  case OWL_HORST =&gt; new ForwardRuleReasonerOWLHorst(spark.sparkContext)
}


// compute inferred graph
val inferredGraph = reasoner.apply(graph)

println(s"|G_inf| = ${inferredGraph.size()}")



------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lorenz Bühmann"<notifications@github.com&gt;; 
发送时间: 2020年12月23日(星期三) 下午4:36
收件人: "SANSA-Stack/SANSA-Stack"<SANSA-Stack@noreply.github.com&gt;; 
抄送: "彭瘦瘦"<2550420029@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [SANSA-Stack/SANSA-Stack] org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to exprId on unresolved object, tree: &#39;rel0.p (#124)





 
您能更具体一点吗？
 
您到底做了什么动作？我们必须查看代码和/或如何运行。
 以及哪个SANSA Stack版本？
 
—
您收到此消息是因为您创建了线程。
直接回复此电子邮件，在GitHub上查看，或取消订阅。

**Comment by LorenzBuehmann** [comment]
does [this](https://github.com/SANSA-Stack/SANSA-Stack/blob/develop/sansa-inference/README.md#from-code) answer your question?

It works with latest version from `develop` version. How to build and use the stack is also documented [here](https://github.com/SANSA-Stack/SANSA-Stack#usage)

**Comment by pj001** [comment]
Are you sure you can perform reasoning? I don't even know how to write the inference rule file






------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lorenz Bühmann"<notifications@github.com&gt;; 
发送时间: 2020年12月23日(星期三) 下午5:08
收件人: "SANSA-Stack/SANSA-Stack"<SANSA-Stack@noreply.github.com&gt;; 
抄送: "彭瘦瘦"<2550420029@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [SANSA-Stack/SANSA-Stack] org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to exprId on unresolved object, tree: &#39;rel0.p (#124)





 
没有这个回答你的问题？
 
它适用于develop版本中的最新版本。这里也记录了如何构建和使用堆栈
 
—
您收到此消息是因为您创建了线程。
直接回复此电子邮件，在GitHub上查看，或取消订阅。

**Comment by pj001** [comment]
How can rules files be put into executive reasoning? I didn't find the relevant cases







------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lorenz Bühmann"<notifications@github.com&gt;; 
发送时间: 2020年12月23日(星期三) 下午5:08
收件人: "SANSA-Stack/SANSA-Stack"<SANSA-Stack@noreply.github.com&gt;; 
抄送: "彭瘦瘦"<2550420029@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [SANSA-Stack/SANSA-Stack] org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to exprId on unresolved object, tree: &#39;rel0.p (#124)





 
没有这个回答你的问题？
 
它适用于develop版本中的最新版本。这里也记录了如何构建和使用堆栈
 
—
您收到此消息是因为您创建了线程。
直接回复此电子邮件，在GitHub上查看，或取消订阅。

**Comment by LorenzBuehmann** [comment]
I cannot follow you - there are predefined rules sets with RDFS and OWL-Horst: as documented here: https://github.com/SANSA-Stack/SANSA-Stack/blob/develop/sansa-inference/README.md#from-code

The supported reasoning profiles are documented here: https://github.com/SANSA-Stack/SANSA-Stack/tree/develop/sansa-inference#supported-reasoning-profiles

Use the `ForwardRuleReasonerRDFS` and it will compute the inferred triples for RDFS. 
Use the `ForwardRuleReasonerOWLHorst` and it will return all inferred triples for OWL Horst profile.
Use the `ForwardRuleReasonerEL` and it will return all inferred triples for a subset of OWL 2 EL profile.

It's not a general purpose rule-engine. We had some experimental implementation but didn't have worked on it in the meantime.

**Comment by pj001** [comment]
Hello, maybe you don't understand what I mean. I mean I want to use Jena to implement rule-based reasoning. But I don't know where to define the rules file rules in the code. I don't know how to execute the inference program.







------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lorenz Bühmann"<notifications@github.com&gt;; 
发送时间: 2020年12月23日(星期三) 下午5:20
收件人: "SANSA-Stack/SANSA-Stack"<SANSA-Stack@noreply.github.com&gt;; 
抄送: "彭瘦瘦"<2550420029@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [SANSA-Stack/SANSA-Stack] org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to exprId on unresolved object, tree: &#39;rel0.p (#124)





 
我无法跟随您-RDFS和OWL-Horst有预定义的规则集：如此处所述：https : //github.com/SANSA-Stack/SANSA-Stack/blob/develop/sansa-inference/README.md#from -码
 
支持的推理配置文件记录在这里：https : //github.com/SANSA-Stack/SANSA-Stack/tree/develop/sansa-inference#supported-reasoning-profiles
 
使用ForwardRuleReasonerRDFS和将计算RDFS的推断三元组。
 使用ForwardRuleReasonerOWLHorst和，它将返回OWL Horst配置文件的所有推断的三元组。
 
—
您收到此消息是因为您创建了线程。
直接回复此电子邮件，在GitHub上查看，或取消订阅。

**Comment by LorenzBuehmann** [comment]
This is not Jena and the rules support in SANSA Stack is not a general purpose rule-engine. We had some experimental implementation but didn't have worked on it in the meantime.

If you simply want to use Jena, use their [`GenericRuleReasoner`](https://jena.apache.org/documentation/javadoc/jena/org/apache/jena/reasoner/rulesys/GenericRuleReasoner.html) - but it's far from being trivial to apply their approach in a distributed settings - at least not efficient (indeed fixpoint iteration and applying each rule could be done, but this can be slow). Their efficiency comes from a RETE based implementation. We do not have this for Spark.

**Comment by pj001** [comment]
Can't I implement rule-based reasoning with spark? Then I don't quite understand what your reasoning means?







------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lorenz Bühmann"<notifications@github.com&gt;; 
发送时间: 2020年12月23日(星期三) 下午5:28
收件人: "SANSA-Stack/SANSA-Stack"<SANSA-Stack@noreply.github.com&gt;; 
抄送: "彭瘦瘦"<2550420029@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [SANSA-Stack/SANSA-Stack] org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to exprId on unresolved object, tree: &#39;rel0.p (#124)





 
This is not Jena and the rules support in SANSA Stack is not a general purpose rule-engine. We had some experimental implementation but didn't have worked on it in the meantime.
 
If you simply want to use Jena, use their GenericRuleReasoner - but it's far from being trivial to apply their approach in a distributed settings. Their efficiency comes from a RETE based implementation. We do not have this for Spark.

**Comment by LorenzBuehmann** [comment]
ok ... then read about RDFS, OWL Horst, OWL EL reasoning - or what do you think this is? I mean, RDFS reasoning just provides a fixed set of entailment rules of the RDFS semantics - computing those inferred triples is reasoning.
Here are the RDFS rules: https://www.w3.org/TR/rdf11-mt/#patterns-of-rdfs-entailment-informative

So why is this **not** reasoning? How do you call this?

**Comment by LorenzBuehmann** [comment]
> Can't I implement rule-based reasoning with spark?

I'm sure you can do it. Just write the implementation for it using Spark Core API and Scala. All I was saying is that in SANSA Stack there is no such generic rule engine - sorry.

**Comment by pj001** [comment]
Can I realize that a is B's father, B is C's father and infer that a is C's grandfather?







------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lorenz Bühmann"<notifications@github.com&gt;; 
发送时间: 2020年12月23日(星期三) 下午5:37
收件人: "SANSA-Stack/SANSA-Stack"<SANSA-Stack@noreply.github.com&gt;; 
抄送: "彭瘦瘦"<2550420029@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [SANSA-Stack/SANSA-Stack] org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to exprId on unresolved object, tree: &#39;rel0.p (#124)





 
ok ... then read about RDFS, OWL Horst, OWL EL reasoning - or what do you think this is? I mean, RDFS reasoning just provides a fixed set of entailment rules of the RDFS semantics - computing those inferred triples is reasoning.
 
So why is this not reasoning? How do you call this?

**Comment by LorenzBuehmann** [comment]
Not with SANSA Stack. As I said before, we do not have a generic rule engine - I'm sorry, but nobody is working on it. 

On the other hand, feel free to contribute - we're always happy to get new features and support from other people.
Just write the Scala or Java code and make use of the Spark Core or SQL API - then open a PR at Github and somebody will review the code

**Comment by pj001** [comment]
Where is the application scenario of this framework? I'm confused.







------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lorenz Bühmann"<notifications@github.com&gt;; 
发送时间: 2020年12月23日(星期三) 下午5:42
收件人: "SANSA-Stack/SANSA-Stack"<SANSA-Stack@noreply.github.com&gt;; 
抄送: "彭瘦瘦"<2550420029@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [SANSA-Stack/SANSA-Stack] org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to exprId on unresolved object, tree: &#39;rel0.p (#124)





 
Not with SANSA Stack. As I said before, we do not have a generic rule engine - I'm sorry, but nobody is working on it.
 
On the other hand, feel free to contribute - we're always happy to get new features and support from other people.
 Just write the Scala or Java code and make use of the Spark Core or SQL API - then open a PR at Github and somebody will review the code

**Comment by LorenzBuehmann** [comment]
of this "framework"? Batch processing of RDF data in Spark?

Having an RDFS or OWL ontology as schema is the starting point for our reasoning. Some schema triples/axioms - that's all.
Then make implicit triples explicit.

Anyways ... if you mean any RDFS or OWL reasoner is useless ... fair enough. You now, sometimes people want to work on the inferred facts, like having the domain `C` of a property `p` meaning that subjects `s_i` in RDF triples `(s_i, p, o)` belong to class `C`. Having `A` being subclass of `B` and knowing individual `i` belongs to `A` we can infer that `i` belongs to `B` as well. But if this isn't enough for you, go for it and implement something more expressive and powerful. 

To sum up, I think we cannot help you with your requirements, we're sorry - hopefully you'll find some other Open Source API or framework that will help you solving your tasks. There are also triple stores with custom rule support if you think this is what you need. Or commercial rule engines like Drools. If all those stuff is still not sufficient for your use-case,  it should be easy for you to implement it by yourself.

Thank you for your questions and making things more clear.

Good luck, take care and Merry Christmas!

**Comment by LorenzBuehmann** [comment]
Answer provided.
